Add Typesense vector store

jasonbosco opened the pull request:
Related to #931.

Starting this PR to collaborate.

jasonbosco committed:
First draft of Typesense integration

dev2049 commented:
will try to pick this back up this week!

jasonbosco commented:
Thank you @dev2049!

dev2049 committed:
merge

dev2049 committed:
merge

dev2049 commented on langchain/vectorstores/typesense.py:
```diff
@@ -0,0 +1,311 @@
+                            ],
+                        }
+                    )
+                except ObjectAlreadyExists:
```
how could we get an ObjectAlreadyExists if we only go down this path after an ObjectNotFound error?

jasonbosco replied:
Hmmm, good point! Added it due to force of habit!

Should be safe to remove the exception handling here.

dev2049 commented on langchain/vectorstores/typesense.py:
```diff
@@ -0,0 +1,311 @@
+                                    "type": "float[]",
+                                    "num_dim": len(docs[0]["vec"]),
+                                },
+                                {"name": "text", "type": "string"},
```
is this name supposed to be text_key?

jasonbosco replied:
Yup, I've updated it

dev2049 commented on langchain/vectorstores/typesense.py:
```diff
@@ -0,0 +1,311 @@
+            embedding = self._embedding_function(text)
+            metadata = metadatas[i] if metadatas else {}
+            docs.append(
+                {"id": ids[i], "vec": embedding, "text": text, "metadata": metadata}
```
is "text" supposed to be text_key?

jasonbosco replied:
Yup, I've updated it

dev2049 committed:
cr

dev2049 commented:
@jasonbosco updated. had a couple outstanding questions and would generally be great to get your eyes on it to make sure the default params make sense. if things generally look ok we can add documentation and tests

jasonbosco committed:
Allow for importing Typesense using `from langchain.vectorstores import Typesense` to be consistent with other vectorstores

jasonbosco committed:
Fix collection name

jasonbosco committed:
Prefix auto-generated collection name with `langchain`

jasonbosco committed:
Fix result parsing

jasonbosco committed:
Show keys explicitly for each of reading code

jasonbosco committed:
Use dynamic text key

jasonbosco commented:
@dev2049 Thank you for the updates! I've addressed your comments in the set of commits I just pushed. 

I also found and fixed a few issues I ran into when I used this branch with an [actual langchain example](https://github.com/typesense/langchain-integration-testground).

jasonbosco committed:
Add docs for Typesense VectorStore

jasonbosco commented:
@dev2049 I just added docs for Typesense similar to the other vector stores: https://github.com/jasonbosco/langchain/blob/85b7c739518b928eebb8224fb82fbaf0c3c43010/docs/modules/indexes/vectorstores/examples/typesense.ipynb

jasonbosco committed:
Lint

jasonbosco committed:
Lint more
